src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,378 @@ namespace xrt {
-
-    xclBOProperties prop;
-    if (xclGetBOProperties(handle, bohdl, &prop)) {
-      syslog(LOG_ERR, "Cannot get SK .so BO info.\n");
-      return -1;
-    }
-
-    buf = xclMapBO(handle, bohdl, false);
-    if (!buf) {
-      syslog(LOG_ERR, "Cannot map softkernel BO.\n");
-      return -1;
-    }
-
-    snprintf(path, XRT_MAX_PATH_LENGTH, "%s", SOFT_KERNEL_FILE_PATH);
-
-    /* If not exist, create the path one by one. */
-    std::filesystem::create_directories(path);
-
-    fptr = fopen(sk_path, "w+b");
-    if (fptr == NULL) {
-      syslog(LOG_ERR, "Cannot create file: %s\n", sk_path);
-      munmap(buf, prop.size);
-      return -1;
-    }
+      xclBOProperties prop = {};
+      int ret = xclGetBOProperties(handle, bohdl, &prop);
+      if (ret) {
+	  const auto errMsg = "Unable to get BO properties!";
+	  xrt_core::message::send(severity_level::error, "SKD", errMsg);
+      }
 
-    /* copy the soft kernel to file */
-    if (fwrite(buf, prop.size, 1, fptr) != 1) {
-      syslog(LOG_ERR, "Fail to write to file %s.\n", sk_path);
-      fclose(fptr);
-      munmap(buf, prop.size);
-      return -1;
-    }
-    syslog(LOG_INFO,"File created at %s\n", sk_path);
+      auto buf = reinterpret_cast<char *>(xclMapBO(handle, bohdl, false));
+      if(buf == MAP_FAILED) {
+	  const auto errMsg = "Cannot map soft kernel BO!";
+	  xrt_core::message::send(severity_level::error, "SKD", errMsg);
+	  return -errno;
+      }
 
-    fclose(fptr);
-    munmap(buf, prop.size);
+      const std::filesystem::path path(SOFT_KERNEL_FILE_PATH+m_xclbin_uuid.to_string()+"/");
+      xrt_core::message::send(severity_level::info, "SKD", path.string());
+
+      if(!std::filesystem::exists(path))
+	  std::filesystem::create_directories(path);
+
+      // Check if file is the same
+      if(!std::filesystem::exists(m_sk_path)) {
+	  std::ofstream fptr(m_sk_path, std::ios::out | std::ios::binary);
+	  if (!fptr.is_open()) {
+	      const auto errMsg = boost::format("Cannot create file: %s") % m_sk_path.string();
+	      xrt_core::message::send(severity_level::error, "SKD", errMsg.str());
+	      xclUnmapBO(handle, bohdl, buf);
+	      return -EPERM;
+	  }
+	  
+	  // copy the soft kernel to file
+	  fptr.write(buf, prop.size);
+	  if (fptr.fail()) {
+	      const auto errMsg = boost::format("Fail to write to file ") % m_sk_path.string();
+	      xrt_core::message::send(severity_level::error, "SKD", errMsg.str());
+	      fptr.close();
+	      xclUnmapBO(handle, bohdl, buf);
+	      return -EIO;
+	  }
+	  const auto msg = boost::format("File created at %s") % m_sk_path.string();
+	  xrt_core::message::send(severity_level::info, "SKD", msg.str());
+	  
+	  fptr.close();
+	  xclUnmapBO(handle, bohdl, buf);
+      }
 
-    return 0;
+      return 0;
   }
 
   /*
    * This function delete the soft kernel file.
    */
-  int skd::deleteSoftKernelFile()
+  int skd::delete_softkernelfile()
   {
-    return(remove(sk_path));
+      if(std::filesystem::exists(m_sk_path))
```

Comment:
Please use `std::filesystem::remove()` and you do not need to do the `exists `check.